Objects/setobject.c

```diff
@@ -1063,36 +1063,73 @@ make_new_set_basetype(PyTypeObject *type, PyObject *iterable)
 /* The empty frozenset is a singleton */
 static PyObject *emptyfrozenset = NULL;
 
+static PyObject *
+make_new_frozenset(PyTypeObject *type, PyObject *iterable)
+{
+    if (iterable != NULL) {
+        /* frozenset(f) is idempotent */
+        if (PyFrozenSet_CheckExact(iterable)) {
+            Py_INCREF(iterable);
+            return iterable;
+        }
+        PyObject *res = make_new_set((PyTypeObject *)type, iterable);
+        if (res == NULL || PySet_GET_SIZE(res)) {
+            return res;
+        }
```

Comment:
This code path is non-obvious. I suggest to add a comment:

```
/* If the created frozenset is empty, return the empty frozenset singleton instead */
```

Maybe also replace `PySet_GET_SIZE(res)` with `PySet_GET_SIZE(res) != 0` to make the intent more obvious.